src/h3lib/lib/edge.c

```diff
@@ -0,0 +1,230 @@
+    H3Index nonNormalizedEdge;
+    H3Error nonNormalizedError =
+        cellsToEdgeNonNormalized(cell1, cell2, &nonNormalizedEdge);
+    if (nonNormalizedError) {
+        return nonNormalizedError;
+    }
+    H3Index normalizedEdge;
+    H3Error normalizeError = normalizeEdge(nonNormalizedEdge, &normalizedEdge);
+    if (normalizeError) {
+        return normalizeError;
+    }
+    *out = normalizedEdge;
+    return E_SUCCESS;
+}
+
+/**
+ * Determines if the provided H3Index is a valid edge index
+ * @param edge The edge H3Index
+ * @return 1 if it is an edge H3Index, otherwise 0.
+ */
+int H3_EXPORT(isValidEdge)(H3Index edge) {
+    if (H3_GET_MODE(edge) != H3_EDGE_MODE) {
+        return 0;
+    }
+    Direction neighborDirection = H3_GET_RESERVED_BITS(edge);
+    if (neighborDirection <= CENTER_DIGIT || neighborDirection >= NUM_DIGITS) {
+        return 0;
+    }
+
+    H3Index cells[2] = {0};
+    // We also rely on the first returned cell being the "owning" cell.
+    H3Error cellsResult = H3_EXPORT(edgeToCells)(edge, cells);
+    if (cellsResult) {
+        return 0;
+    }
+    if (H3_EXPORT(isPentagon)(cells[0]) && neighborDirection == K_AXES_DIGIT) {
+        // Deleted direction from a pentagon
+        return 0;
+    }
+    if (cells[1] < cells[0]) {
+        // Not normalized
+        return 0;
+    }
+
+    return H3_EXPORT(isValidCell)(cells[0]);
+}
+
+/**
+ * Returns the cell pair of hexagon IDs for the given edge ID.
+ *
+ * The first cell returned is always the "owning" cell of the edge.
+ * @param edge The edge H3Index
+ * @param cells Pointer to memory to store cell IDs
+ */
+H3Error H3_EXPORT(edgeToCells)(H3Index edge, H3Index *cells) {
+    // Note: this function will accept directed edges as well, but report
+    // E_UNDIR_EDGE_INVALID errors.
+    H3Index directedEdge = edgeAsDirectedEdge(edge);
+    H3Error cellsResult = H3_EXPORT(directedEdgeToCells)(directedEdge, cells);
+    if (cellsResult) {
+        return wrapDirectedEdgeError(cellsResult);
+    }
+    return E_SUCCESS;
+}
+
+/**
+ * Provides all of the edges from the current H3Index.
+ * @param origin The origin hexagon H3Index to find edges for.
+ * @param edges The memory to store all of the edges inside.
+ */
+H3Error H3_EXPORT(cellToEdges)(H3Index origin, H3Index *edges) {
+    // Determine if the origin is a pentagon and special treatment needed.
+    int isPent = H3_EXPORT(isPentagon)(origin);
+
+    // This is actually quite simple. Just modify the bits of the origin
+    // slightly for each direction, except the 'k' direction in pentagons,
+    // which is zeroed.
+    for (int i = 0; i < 6; i++) {
+        if (isPent && i == 0) {
+            edges[i] = H3_NULL;
+        } else {
+            H3Index edge = origin;
+            H3_SET_MODE(edge, H3_EDGE_MODE);
+            H3_SET_RESERVED_BITS(edge, i + 1);
+            H3Error normalizeError = normalizeEdge(edge, &edges[i]);
```

Comment:
Nit: The same perf concerns apply here. Every edge here is potentially non-normalized, so we could end up running `directionForNeighbor` 6 times. I think it might be more efficient to take `gridDisk(origin, 1)` and then loop through the neighbors running `cellsToEdge` for each. You could even use `edges` for the storage and then replace each neighbor with the origin-neighbor edge, that way you don't have to allocate any new memory.

Reply:
`edges` can't be used for storage because gridDisk needs an array of length 7. gridRing can't be used because there is no "safe" version. We would need a different function for getting just the immediate neighbors of a cell.